foundation-model-stack/fms-hf-tuning
# Install Acceleration Framework into Training Script

### Description of the change

This PR installs the Acceleration Framework into the `sft_trainer.py`. This is a followup to #119 which proposes to have a lightweight integration; the implementation of Acceleration Framework is kept seperate in the repo [fms-acceleration](https://github.com/foundation-model-stack/fms-acceleration) under the same `foundation-model-stack` org.
- introduce a `AccelerationFrameworkArguments` that accepts an `--acceleration_framework_config_file` argument to configure the framework.
- update `pyproject.toml` with an *optional dependency* `[fms-accel]` that installs the [fms-acceleration framework](https://github.com/foundation-model-stack/fms-acceleration/plugins/framework).
- update `README.md` to include basic usage of `fms-acceleration`, using it to perform accelerated PEFT with a 4bit GPTQ-LoRA.
- ensure that the integration within `sft_trainer.py` is optional, the `.get_framework` call  below will **silently disable** framework if `fms-accel` dependency is not desired.
    ```
    framework = AccelerationFrameworkConfig.from_dataclasses(*dataclass_configs).get_framework()
    ```


- restrict to only *three integration points* within `sft_trainer.py` script:
  1. `framework.model_loader`: load model if `framework.requires_custom_loading == True` 
  2. `framework.augmentation`: load if `framework.requires_agumentation == True`
  3. `framework.get_callbacks_and_ready_for_train`: get callbacks and do final prep on `model` and `trainer.accelerator` if required.

**Update: Picture has been changed to reflect new dataclass arguments flow**

![image](https://github.com/foundation-model-stack/fms-hf-tuning/assets/8325951/c612eaa3-2707-4534-8288-d71802cf3686)






### Related issue number

Related to the merged PR #119 for the ADR of Training Acceleration.



### How to verify the PR

This PR can be verified in the following ways:
1. run the new folder `tests/acceleration`, see **Note-for-unit-tests** below.
2. **[Update: this bench needs to be reworked due to YAML arguments being disabled, right now works with some patching]** run the [provided benchmark utility](https://github.com/foundation-model-stack/fms-acceleration/blob/main/scripts/benchmarks/README.md), and checking with the results in the [PR](https://github.com/foundation-model-stack/fms-acceleration/pull/4) and the [a100_80gb](https://github.com/foundation-model-stack/fms-acceleration/blob/main/scripts/benchmarks/refs/a100_80gb.csv) 
   -  consult [this helm chart as a reference](https://gist.github.ibm.com/Aaron-Chew1/f56a53b14849d8c508f35592b05b98c5) to run the bench

**Note for unit tests**
For `tests/acceleration`, note that `testenv` will not install `.fms-accel` dependencies, so some of the newly added tests wil be skipped. So to run *all* the new tests:
```
pip install ".[fms-accel]"
python -m fms_acceleration.cli install peft
pytest tests/acceleration
```



### Was the PR tested
